model/modeldecoder/v2/simdjson.go

```diff
@@ -0,0 +1,593 @@
+
+	"github.com/minio/simdjson-go"
+
+	"github.com/elastic/apm-server/decoder"
+)
+
+func init() {
+	elemsPool.New = func() interface{} {
+		elems := simdjson.Elements{
+			Elements: make([]simdjson.Element, 0, 5),
+			Index:    make(map[string]int, 5),
+		}
+		return &elems
+	}
+	objPool.New = func() interface{} {
+		return &simdjson.Object{}
+	}
+	parsedJSONPool.New = func() interface{} {
+		return &simdjson.ParsedJson{}
+	}
+	iterPool.New = func() interface{} {
+		return &simdjson.Iter{}
+	}
+}
+
+var elemsPool, iterPool, objPool, parsedJSONPool sync.Pool
+
+func simdjsonDecodeEmbeddedMetadata(d decoder.Decoder, out *Metadata) error {
+	return simdjsonDecodeWrappedErr(d, "", out)
+}
+func simdjsonDecodeNestedMetadata(d decoder.Decoder, out *Metadata) error {
+	return simdjsonDecodeWrappedErr(d, "metadata", out)
+}
+
+func simdjsonDecodeWrappedErr(d decoder.Decoder, rootKey string, out *Metadata) error {
+	if err := simdjsonDecode(d, rootKey, out); err != nil {
+		if err == decoder.ErrLineTooLong {
+			return err
+		}
+		return decoder.JSONDecodeError("data read error: " + err.Error())
+	}
+	return nil
+}
+
+func simdjsonDecode(d decoder.Decoder, rootKey string, out interface{}) error {
+	line, err := d.Read()
+	if err != nil {
+		return err
+	}
+	reusedJSON := parsedJSONPool.Get().(*simdjson.ParsedJson)
+	defer parsedJSONPool.Put(reusedJSON)
+	parsedJSON, err := simdjson.Parse(line, reusedJSON)
+	defer parsedJSONPool.Put(parsedJSON)
+	if err != nil {
+		return err
+	}
+	obj := objPool.Get().(*simdjson.Object)
+	defer objPool.Put(obj)
+	tmpIter := iterPool.Get().(*simdjson.Iter)
+	defer iterPool.Put(tmpIter)
+	if err := rootObj(parsedJSON.Iter(), tmpIter, obj); err != nil {
+		return err
+	}
+	if rootKey != "" {
+		if err := keyObject(obj, rootKey); err != nil {
+			return err
+		}
+	}
+
+	if err := decodeElements(obj, "", out); err != nil {
+		return err
+	}
+	return nil
+}
+
+func rootObj(iter simdjson.Iter, tmpIter *simdjson.Iter, obj *simdjson.Object) error {
+	iter.Advance()
+	typ, _, err := iter.Root(tmpIter)
+	if err != nil {
+		return err
+	}
+	if typ != simdjson.TypeObject {
+		err = errors.New("element not an object")
+		return err
+	}
+	obj, err = tmpIter.Object(obj)
+	if obj == nil {
+		return errors.New("object not found")
+	}
+	return err
+}
+
+func keyObject(obj *simdjson.Object, key string) error {
+	elem := obj.FindKey(key, nil)
+	if elem == nil || elem.Type != simdjson.TypeObject {
+		return fmt.Errorf("object with key %v not found", key)
+	}
+	obj, err := elem.Iter.Object(obj)
+	if err != nil {
+		return err
+	}
+	if obj == nil {
+		return errors.New("object not found")
+	}
+	return err
+}
+
+func decodeElements(obj *simdjson.Object, key string, out interface{}) error {
+	elems := elemsPool.Get().(*simdjson.Elements)
+	defer elemsPool.Put(elems)
+	_, err := obj.Parse(elems)
+	if err != nil {
+		return err
+	}
+	for _, elem := range elems.Elements {
+		s := elem.Name
+		if key != "" {
+			s = fmt.Sprintf("%s.%s", key, s)
+		}
```

Comment:
Like in the validation code, it would be great if we could use our knowledge of the schema to generate constants (or perfect hash for constant key lookup), to avoid these allocations.